src/modules/simulator/simulator_mavlink.cpp

```diff
@@ -220,51 +220,47 @@ static void fill_rc_input_msg(input_rc_s *rc, mavlink_rc_channels_t *rc_channels
 	rc->values[17] = rc_channels->chan18_raw;
 }
 
-void Simulator::update_sensors(const mavlink_hil_sensor_t *imu)
+void Simulator::update_sensors(const hrt_abstime &time, const mavlink_hil_sensor_t &imu)
 {
-	// write sensor data to memory so that drivers can copy data from there
-	RawMPUData mpu = {};
-	mpu.accel_x = imu->xacc;
-	mpu.accel_y = imu->yacc;
-	mpu.accel_z = imu->zacc;
-	mpu.temp = imu->temperature;
-	mpu.gyro_x = imu->xgyro;
-	mpu.gyro_y = imu->ygyro;
-	mpu.gyro_z = imu->zgyro;
-
-	write_MPU_data(&mpu);
-	perf_begin(_perf_mpu);
-
-	RawAccelData accel = {};
-	accel.x = imu->xacc;
-	accel.y = imu->yacc;
-	accel.z = imu->zacc;
-
-	write_accel_data(&accel);
-	perf_begin(_perf_accel);
-
-	RawMagData mag = {};
-	mag.x = imu->xmag;
-	mag.y = imu->ymag;
-	mag.z = imu->zmag;
+	if ((imu.fields_updated & 0x1FFF) != 0x1FFF) {
+		PX4_DEBUG("All sensor fields in mavlink HIL_SENSOR packet not updated.  Got %08x", imu.fields_updated);
+	}
 
-	write_mag_data(&mag);
-	perf_begin(_perf_mag);
+	// gyro
+	{
+		float scaling = 1000.0f;
+		_px4_gyro.set_scale(1 / scaling);
+		_px4_gyro.set_temperature(imu.temperature);
+		_px4_gyro.update(time, imu.xgyro * scaling, imu.ygyro * scaling, imu.zgyro * scaling);
+	}
 
-	RawBaroData baro = {};
+	// accel
+	{
+		float scaling = 1000.0f;
+		_px4_accel.set_scale(1 / scaling);
+		_px4_accel.set_temperature(imu.temperature);
+		_px4_accel.update(time, imu.xacc * scaling, imu.yacc * scaling, imu.zacc * scaling);
+	}
 
-	// Get air pressure and pressure altitude
-	// valid for troposphere (below 11km AMSL)
-	baro.pressure = imu->abs_pressure;
-	baro.temperature = imu->temperature;
+	// magnetometer
+	{
+		float scaling = 1000.0f;
+		_px4_mag.set_scale(1 / scaling);
+		_px4_mag.set_temperature(imu.temperature);
+		_px4_mag.update(time, imu.xmag * scaling, imu.ymag * scaling, imu.zmag * scaling);
+	}
 
-	write_baro_data(&baro);
+	// baro
+	{
+		_px4_baro.set_temperature(imu.temperature);
+		_px4_baro.update(time, imu.abs_pressure);
+	}
 
 	RawAirspeedData airspeed = {};
-	airspeed.temperature = imu->temperature;
+	airspeed.temperature = imu.temperature;
 	// FIXME: diff_pressure needs some noise to pass preflight checks, so we just take the
 	//        noise from the gyro.
-	airspeed.diff_pressure = imu->diff_pressure + ((imu->ygyro > 0) ? 0.001f : 0.0f);
+	airspeed.diff_pressure = imu.diff_pressure + ((imu.ygyro > 0) ? 0.001f : 0.0f);
```

Comment:
I had to revert because SITL was now flagging airspeed data as stuck. https://github.com/PX4/Firmware/commit/2142459027ffee7e9a18d466fcd4f8cd0a9613d2#diff-aef1cbcb8457549c454ec4445547f66b

Reply:
No, then something with the noise in sitl_gazebo is wrong!

Reply:
Anyway, I can fix that independently.